src/cell_group_factory.cpp

```diff
@@ -14,26 +16,29 @@ namespace mc {
 using gpu_fvm_cell = mc_cell_group<fvm::fvm_multicell<gpu::backend>>;
 using mc_fvm_cell = mc_cell_group<fvm::fvm_multicell<multicore::backend>>;
 
-cell_group_ptr cell_group_factory(
-        cell_kind kind,
-        cell_gid_type first_gid,
-        const std::vector<util::unique_any>& cell_descriptions,
-        backend_policy backend)
-{
-    switch (kind) {
+cell_group_ptr cell_group_factory(const recipe& rec, const group_description& group) {
+    // Make a list of all the cell descriptions to be forwarded
+    // to the appropriate cell_group constructor.
+    std::vector<util::unique_any> descriptions;
+    descriptions.reserve(group.gids.size());
+    for (auto gid: group.gids) {
+        descriptions.push_back(rec.get_cell_description(gid));
+    }
+
+    switch (group.kind) {
```

Comment:
Performance of the factory was not considered when writing this code. It is only used when setting up the model, and the cost of making the decisions inside the factory are very small relative to the construction costs of the cell groups themselves.

As @halfflat points out, this is more like a placeholder, waiting for support for more more flexible user-defined` cell_group`s.

Note that by the time the `cell_group_factory` is invoked, the decision about which hardware back end the `cell_group` will run on will have been made by the load balancing/domain decomposition methods.